client-common/src/tendermint/rpc_client/sync_rpc_client.rs

```diff
@@ -0,0 +1,287 @@
+use std::{
+    sync::{mpsc::sync_channel, Arc},
+    time::{Duration, SystemTime},
+};
+
+use itertools::izip;
+use serde::Deserialize;
+use serde_json::{json, Value};
+use tendermint::{lite, validator};
+use tokio::runtime::Runtime;
+
+use chain_core::state::ChainState;
+
+use super::async_rpc_client::AsyncRpcClient;
+use crate::{
+    tendermint::{lite::TrustedState, types::*, Client},
+    Error, ErrorKind, Result, ResultExt,
+};
+
+/// Wraps asynchronous RPC client and executes it in tokio runtime
+#[derive(Clone)]
+pub struct SyncRpcClient {
+    runtime: Arc<Runtime>,
+    async_rpc_client: AsyncRpcClient,
+}
+
+impl SyncRpcClient {
+    /// Creates a new synchronous websocket RPC client
+    pub fn new(url: &str) -> Result<Self> {
+        let mut runtime = Runtime::new().chain(|| {
+            (
+                ErrorKind::TendermintRpcError,
+                "Unable to start tokio runtime",
+            )
+        })?;
+
+        let async_rpc_client = runtime
+            .block_on(async { AsyncRpcClient::new(url).await })
+            .chain(|| {
+                (
+                    ErrorKind::TendermintRpcError,
+                    format!("Unable to connect to tendermint RPC websocket at: {}", url),
+                )
+            })?;
+
+        Ok(Self {
+            runtime: Arc::new(runtime),
+            async_rpc_client,
+        })
+    }
+
+    /// Makes an RPC call and deserializes response
+    pub fn call<T>(&self, method: &'static str, params: Vec<Value>) -> Result<T>
+    where
+        T: Send + 'static,
+        for<'de> T: Deserialize<'de>,
+    {
+        let (sender, receiver) = sync_channel(1);
+        let async_rpc_client = self.async_rpc_client.clone();
+
+        self.runtime.spawn(async move {
```

Comment:
should this use `spawn` and not `block_on` or something like that? I think for `spawn`, if some mid-step panics, it may keep hanging?

Reply:
Ideally we should use something like `block_on` here, but not `block_on`, because `block_on` takes mutable reference of `runtime` that means that we cannot `block_on` two different things at the same time (in our case, that'll mean that we cannot make two requests in parallel). The current implementation is mimicking `block_on` with non-mutable reference of `runtime`.

Reply:
what about `block_on_all`?

Reply:
I don't know of any `block_on_all` function. Can you post docs link here?

Reply:
oh, it's 0.2

Reply:
if you do say `panic!("test")` after response, the whole process will fail, right?

Reply:
I've added a `recv_timeout` when reading from response channel. That should be enough?

Reply:
did you try with that `panic!`?

Reply:
Where do you want to put `panic!`?

Reply:
I'm a bit confused. Are you worried about `async` code panicking or are you worried about it getting stuck when it panics. After `recv_timeout` change, if the `async` code panics, user will see a `"Request timed out"` error.

Reply:
e.g. between lines 64 and 65

Reply:
As I mentioned earlier, if `async` code panics, user will see `"Request timed out"` error.

Reply:
so the process will keep hanging even though it's not recoverable?
that's not ideal, but if there's no way around it

Reply:
I'm still not able to understand how the async code can panic? There is not a single `unwrap` or `expect` in async code. If you're worried about someone making changes in future and putting an `unwrap` in async code (which ideally shouldn't be done and we should use proper error handling), then yes, user will keep getting this error when internal code panics.

Reply:
it's not only the direct code here, but all the dependencies -- the websocket library, tendermint-rs, parsers... 
if the process exited, it'd be faster to find out it went wrong because of this (rather than diagnosing tendermint startup etc.)

Reply:
Yeah. This should ideally be done with panic-hooks. I'll investigate about this and create a new PR with improvements (if there are some).